### DISCUSSION TITLE
Calling all SIG chairs/members! (Feature Grid/Roadmap transparency discussion)

### DISCUSSION THREAD

**Discussion by jeremyong-az** [o3de/o3de#6703] [Announcements]
Hey O3DE community,

The TSC would like to maintain a living document that provides additional transparency about the state of the various systems and subsystems in the engine. In the interest of eventually converging on a format we can agree on, Here are some items we should discuss:

## What is being proposed?
  * The sigs collectively maintain a table providing information about engine features
  * The table should indicate the relative maturity of a feature, any known limitations, deprecation plans, or upgrade plans
  * Users reading the table should be able to do a coarse evaluation to determine whether the engine suits their needs, and/or understand what work is needed to fill any gaps
  * The table should be granular enough to be useful, but not so granular that the table is unmaintainable
## What does an entry in the table look like?
  * _Note: this is a proposed starting point_
  * **Sig** (e.g. sig-graphics-audio)
  * **Feature** (e.g. spot lights, culling)
  * **Code status** (e.g. one of `unknown`, `stable`, `volatile`, `deprecated`)
    * Unknown means the code is working but has had insufficient use to really harden any corner cases or identify potential pathological issues
    * Stable means the code is actively used without any serious impediments and has no immediate upgrades planned
    * Volatile means the code is in a heavy state of flux (API and/or data stability is not guaranteed)
    * Deprecated means the code is slated to go away (accompanied with some indication of what the system will be replaced by)
  * **Limitations** (e.g. scalability limits, platform limits, missing features, interoperability issues with content and/or other features)
  * **Comments** (any additional information to provide context about other data in the field)

Please reply with a comment if you have suggestions/feedback about any of the above. Also, please offer suggestions on where you would like such a doc to be hosted.

**Comment by sptramer** [comment]
I would further suggest that `Code status` be broken down into a few layers such as "API" and "Implementation" at minimum (a volatile feature can have a stable API, and a deprecated API can be in front of a stable and undeprecated feature!) as well as some form of "Docs coverage" information.

I should also mention that this information would be best broken down by Framework and Gem, to avoid an overwhelmingly large table of information. This would also allow for feature status to be updated based on Gem versioning independent of O3DE versioning.

I can also see the implications of `volatile` and `deprecated` on documentation - it's very likely that sig-docs-community would argue against documenting (or keeping documentation for) features in these types of states, and we would need strategies for organizing information around highly volatile features.

**Comment by jeremyong-az** [comment]
Other suggestions received out-of-band:

- Relabeling "unknown" as "untested" or "unproven" for clarity
- An explicit "needs refactor" or "needs re-architecture" label

**Comment by OBWANDO** [comment]
Example for one System on its own wiki/github page that is published for discussion:

**Graphics Rendering Subsystem**

| Subsystem | Detail | Feature    | Code | Release | Link | Doc | 
|:------|:---------|:--------|:-------|:------------|:-----------:|:-------:|
|   Rendering |           |        |         |        |         |          | 
|           | Directx12 |           |        |         |  [RFC/Detail](https://github.com/o3de/o3de/issues/1847)      |   [Link](https://o3de.org/docs/atom-guide/features/) | 
|           |   Vulkan     |  🟡 Partial  |  🟢 Stable      |  🟠 Volatile  |        |  [Link](https://o3de.org/docs/atom-guide/features/) | 
|           |   Metal     |   🟡 Partial      |  🔴 Unstable       |  🔴 Untested       |       |   [Link](https://o3de.org/docs/atom-guide/features/) | 
|           |   UberRender     |   🟠 Basic |  🟠 Volatile       |  ⭕ Unknown|        |    [Link](https://o3de.org/docs/atom-guide/features/) | 
|    |  Nanite-ish         |  ⭕ In Research      |  ❌ Not Started    |   ❌ Not Started     |         |          | 
|   Raytracing     |           |        |         |        |         |          | 
|           | Directx12 |           |        |         |        |  [Link](https://o3de.org/docs/atom-guide/features/) | 
|           |   Vulkan     |           |        |         |        |  [Link](https://o3de.org/docs/atom-guide/features/) | 
|           |   Metal     |           |        |         |        |   [Link](https://o3de.org/docs/atom-guide/features/) | 
|   Anti-Aliasing     |           |        |         |        |         |   [Link](https://o3de.org/docs/atom-guide/features/) | 
|        |  MSAA         |        |         |        |         |          | 
|        |  SMAA        |        |         |        |         |          | 
|        |  TAA         |        |         |        |         |          | 
|   Shaders     |           |        |         |        |         |          | 
|        |  HLSL         |        |         |        |         |          | 
|        |  AZSL         |        |         |        |         |          | 
|        |  OLDZSL         |  🟢 Complete       | 🟤 Deprecate |  🟢 Released       |          |          | 
|   Global Illumination     |           |        |         |        |         |          | 
|        | Forward+ |        |         |        |         |          | 
|        | Deferred |        |         |        |         |          | 
|        | Hybrid |        |         |        |         |          | 
|   Lighting     |           |        |         |        |         |         |  
|        |  Directional         |        |         |        |         |          | 
|        |  Point         |        |         |        |         |          | 
|        |  Global Sky        |        |         |        |         |          | 
|        |  Area        |        |         |        |         |          | 
|        |  Forward shading         |        |         |        |         |          | 
|        |  Diffuse Probe Grid         |        |         |        |         |          | 
|        |  Reflection Probes    |        |         |        |         |          | 
|        |  Subsurface Scattering         |        |         |        |         |          | 
|        |  Exposure / Eye Adaptation |        |         |        |         |          | 
|        |  HDRI Skybox |        |         |        |         |          | 
|        |  Decal |        |         |        |         |          | 
|  PostFX   |   |        |         |        |         |          | 
|        | Tonemapping          |        |         |        |         |          | 
|        | Color Grading          |        |         |        |         |          | 
|        | Post-Process Volumes     |        |         |        |         |          | 
|        | Bloom          |        |         |        |         |          | 
|        | Deferred Fog          |        |         |        |         |          | 
|        | Depth of Field          |        |         |        |         |          | 
|        | SSAO          |        |         |        |         |          | 

Unicode Colors and objects that work in Github Markdown
🔴🟠🟡🟢🔵🟣🟤⚫⚪🔘🛑⭕❌
🟥🟧🟨🟩🟦🟪🟫⬛⬜🔲🔳⏹☑✅❎
❤️🧡💛💚💜💙🤎🖤🤍♥️💔💖💘💝💗💓💟💕❣️♡
🔺🔻🔷🔶🔹🔸♦💠💎💧🧊
🏴🏳🚩🏁
◻️◼️◾️◽️▪️▫️
https://unicode-table.com/en/emoji/symbols/

**Reply by jeremyong-az** [reply]
Great start!

Column suggestions:

| Subsystem | Detail | Feature    | Content | Function | API | Performance | Link | Doc | 
|:------|:---------|:--------|:-------|:------------|:-----------:|:-------:|:--:|:--:|

Content: content stability (is content built with this subsystem/feature expected to change in the near future)
Function: is the subsystem feature-complete, bug-free, and functionally stable
API: are user-facing APIs (public headers, interfaces) in a state of flux or stable

I suggest keeping a live-at-head version of the table representing the state of the development branch, and a static version per point-release.

**Reply by jeremyong-az** [reply]
Suggested JSON representation (which can be used to generate markdown tables or documentation). Instead of unicode, the various stability/perf scores are given a simple 0 to 3 inclusive range (which can be converted to unicode or some other representation in the future)

```jsonc
{
    "Version": "development", // Each time the development branch is cut, this version field is set to the engine version about to be published
    "SIG": "sig-graphics-audio", // Each SIG maintains their own stability doc and all docs are merged and published each release
    "Subsystems": {
        // Subsystem entries are keyed to the subsystem name
        "Render Hardware Interface": {
            // Entries within a subsystem are keyed to the feature name
            "Sparse Textures": {
                "MacOS": false, // By default, features are supported on all platforms. Adding a platform and setting its value to false indicates limited or no support on a given platform
                "Description": "Brief note describing what this does",
                "Content": 0, // Numerical score indicating stability (0 = unsupported, 1 = deprecated, 2 = experimental/volatile, 3 = stable
                "Functional": 0, // Numerical score indicating stability (0 = unsupported, 1 = deprecated, 2 = experimental/volatile, 3 = stable
                "Code/API": 0, // Numerical score indicating stability (0 = unsupported, 1 = deprecated, 2 = experimental/volatile, 3 = stable
                "Performance": 0, // Numerical score indicating performance hardening (0 = feature unsupported, 1 = feature requires performance testing, 2 = feature requires optimizations to ship , 3 = feature deemed sufficiently optimized, with scalability measurements in notes),
                "Link": "url link to additional information",
                "Doc": "documentation URL for usage notes"
            },
            // Other subsystem features
        },
        // Other subsystems
    }
}
```

**Reply by sptramer** [reply]
Docs might want to provide an alternate structure here (for extensibility purposes), so `"Doc": { /* TBD */ }`. We've got a lot of information we could throw up per-sig or per-module if we can get RGY coverage dashboards working.

**Comment by FiniteStateGit** [comment]
I could see scripting support as a separate category to be tracked. An API could be both complete and documented, but at the same time unreflected.

**Reply by nick-l-o3de** [reply]
The documentation around this feature and process should keep the audience in mind.  The audience (argue here!) being people who are considering this engine for a project, or are considering becoming maintainers or ingesting the engine technology into their engine stack and want to understand the current state of various features that may be relevant to their use case.

Most specifically, its not a list of planned features, or roadmap items.   It _can_ list features that do not exist in the engine (And note that they are unsupported) if the information would be relevant to someone considering the engine.  This means the granularity is decided by each SIG and should be streamlined to convey this information to would-be engine adopters and evaluators.

**Comment by lmbr-pip** [comment]
If we are moving towards producing raw JSON or other data format, we should provide some guidance/formatting tools to make it easy to review changes. I imagine SIGs will have 10s/100s features to review and we need to make the iteration cycle somewhat efficient (review, edit), SIGs should review with their community to capture open action items/questions and documentation needs that this will invariably generate and seeing the final visual format will really aid this process.

**Reply by OBWANDO** [reply]
You can find the first rev here:
https://o3de.github.io/community/features/form.html
https://github.com/o3de/community/tree/main/features

The code runs right from the GitHub repo when published as a page on github.io

The code isn't pretty, it's a rough frame up, but can generate the sheets from JSON,  manipulate the data, and export/ download the JSON changes to a file.
